core/src/test/java/io/cucumber/core/logging/WithLogRecordListener.java

```diff
@@ -0,0 +1,48 @@
+package io.cucumber.core.logging;
+
+import io.cucumber.core.logging.LogRecordListener;
+import io.cucumber.core.logging.LoggerFactory;
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ ElementType.TYPE, ElementType.METHOD })
+@Retention(RetentionPolicy.RUNTIME)
+@ExtendWith({ WithLogRecordListener.Extension.class })
+public @interface WithLogRecordListener {
+    class Extension implements BeforeEachCallback, AfterEachCallback, ParameterResolver {
+        protected LogRecordListener logRecordListener;
```

Comment:
Because tests maybe run in parallel with JUnit, it's generally good to keep extensions stateless and store any state in the context.